### ISSUE TITLE
[java] Deprecate rule CloneThrowsCloneNotSupportedException

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#3112]
[CloneThrowsCloneNotSupportedException](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#clonethrowsclonenotsupportedexception)

From #3104

>  I'm not sure CloneThrowsCloneNotSupportedException is useful, as calling super.clone() already forces you to declare the exception or handle it... why does it exist?

> Don't know - that's a very old rule. Object.clone() throws this checked Exception, but you don't need to declare that when you override the clone method. You might need to catch it though when calling super.clone(). Maybe a candidate for retirement?

**Comment by adangel** [comment]
Note that this rule is also currently on the quickstart ruleset.

While looking through the test cases, I found one reason, why this rule might be useful:

Suppose you have a base class like this, that does not throw CloneNotSupportedException:

```java
public class Base {
    @Override
    protected Base clone() {
        try {
            return (Base) super.clone();
        } catch (CloneNotSupportedException e) {
            throw new RuntimeException(e);
        }
    }
}
```

Then you have no chance to throw that exception in the subclass:

```java
public class Sub extends Base {
    @Override
    protected Sub clone() throws CloneNotSupportedException { // compile error here: Exception CloneNotSupportedException is not compatible with throws clause in Base.clone()
        throw new CloneNotSupportedException(); // next compile error here: Unhandled exception type CloneNotSupportedException
    }
}
```

The rule currently ignores final classes, since there you can't have subclasses, so you can't run into this situation.

We could argue though, that it's a design decision if you remove the throws clause in a base class: This gives a hint, that all subclasses must support cloning....

As it can be seen in the Base class:

> as calling super.clone() already forces you to declare the exception or handle it... 

The compiler takes care of all that already. Either you need to rethrow (via throws clause) or you need to catch and handle it.

In the above example, if you change Sub to the following, so that is compiles:

```java
public class Sub extends Base {

    @Override
    protected Sub clone() /*throws CloneNotSupportedException*/ {
        //throw new CloneNotSupportedException();
        return (Sub) super.clone();
    }
}
```

you'll still get a "CloneNotSupportedException" wrapped in a "RuntimeException" when you try to execute `new Sub().clone()`, because Sub does not implement the `Cloneable` interface.

In conclusion, I still think, that this rule has no real value. As you can see in the Base example class, you are forced to deal with that exception in one way or the other. Instead of throwing you could `return null` here - then you don't get an exception, but you'd need to deal with null in your program. At some point, you'd need to think about what to do, if clone is not supported. And always declaring CloneNotSupportedException is only one way to deal with it.
